## STEP 3 Part 1 of 2 - FINAL PROJECT
### Part 1: Find best model
### Rakamin - Data Science Batch 47
## Group 3 <br>
# **Byte Me** <br>
Hijir Della Wirasti <br>
Mauliddinia Iftikhar Agnany <br>
Jericho Medion Haryono <br>
Fakhri Dwi Nugroho <br>
Ryan Nofandi <br>

Laporan: https://docs.google.com/document/d/1FLA2xAlc_K8GkiqfJ7efhuxjzPcuUDJpxs5HS1Bn9i8/edit?usp=sharing

Dataset awal: Home Credit Default Risk<br>
https://www.kaggle.com/competitions/home-credit-default-risk/data <br>

Dataset setelah Pre-processing<br>
https://drive.google.com/file/d/1cX9JxNyOKYMBXlVw2ewd3HxzwsbnGnbL/view?usp=sharing



In [ ]:
import pandas as pd
import numpy as np
from scipy import stats

import seaborn as sns
import matplotlib.pyplot as plt

from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, ConfusionMatrixDisplay
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler
import warnings

# Ignore all warnings
warnings.filterwarnings("ignore")


## Read the Data

In [ ]:
df_train_log = pd.read_csv('df_train_log.csv')

# Feature Scaling

In [ ]:
from sklearn.preprocessing import StandardScaler

# Memisahkan fitur (X) dan target (y)
X = df_train_log.drop(columns='TARGET')
y = df_train_log['TARGET']

# Inisialisasi StandardScaler
scaler = StandardScaler()

# Terapkan StandardScaler ke fitur
X_scaled = scaler.fit_transform(X)

# Menampilkan beberapa baris pertama dari X_scaled
print(pd.DataFrame(X_scaled, columns=X.columns).head())

   NAME_CONTRACT_TYPE  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  \
0            0.324189     1.390300     -0.715531         0.665112   
1            0.324189    -0.719269     -0.715531        -1.503506   
2           -3.084622     1.390300      1.397564         0.665112   
3            0.324189    -0.719269     -0.715531         0.665112   
4            0.324189     1.390300     -0.715531         0.665112   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0     -0.589199          0.633425   -0.215579     0.086951   
1     -0.589199          1.222404    1.402158     0.761529   
2     -0.589199         -1.615780   -1.756794    -2.289051   
3     -0.589199         -0.196692   -0.582710     0.423738   
4     -0.589199         -0.412398    0.109361    -0.136354   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  ...  \
0                         0.0    1.509513  ...   
1                         0.0   -0.158701  ...   
2                         0.0   -0.679675  ...   
3           

# Handle Class Imbalance

## Oversampling (SMOTE)

In [ ]:
from imblearn.over_sampling import SMOTE
from collections import Counter

# Inisialisasi SMOTE
smote = SMOTE(sampling_strategy=0.5)  # Menambah kelas minoritas hingga mencapai 50% dari kelas mayoritas

# Terapkan SMOTE ke data yang sudah discaling
X_smote, y_smote = smote.fit_resample(X_scaled, y)

# Print distribusi kelas setelah SMOTE
print(f'Distribusi kelas sebelum SMOTE: {Counter(y)}')
print(f'Distribusi kelas setelah SMOTE: {Counter(y_smote)}')

# Optional: Tampilkan beberapa baris data hasil resampling
print(pd.DataFrame(X_smote, columns=X.columns).head())


Distribusi kelas sebelum SMOTE: Counter({0: 278843, 1: 24396})
Distribusi kelas setelah SMOTE: Counter({0: 278843, 1: 139421})
   NAME_CONTRACT_TYPE  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  \
0            0.324189     1.390300     -0.715531         0.665112   
1            0.324189    -0.719269     -0.715531        -1.503506   
2           -3.084622     1.390300      1.397564         0.665112   
3            0.324189    -0.719269     -0.715531         0.665112   
4            0.324189     1.390300     -0.715531         0.665112   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0     -0.589199          0.633425   -0.215579     0.086951   
1     -0.589199          1.222404    1.402158     0.761529   
2     -0.589199         -1.615780   -1.756794    -2.289051   
3     -0.589199         -0.196692   -0.582710     0.423738   
4     -0.589199         -0.412398    0.109361    -0.136354   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  ...  \
0                         0.0    1.

## Undersampling

In [ ]:
from imblearn.under_sampling import RandomUnderSampler
from collections import Counter

# Inisialisasi RandomUnderSampler
undersample = RandomUnderSampler(sampling_strategy=0.8)  # Mengurangi jumlah kelas mayoritas hingga mencapai 80% dari jumlah kelas minoritas

# Terapkan RandomUnderSampler ke data yang sudah discaling
X_under, y_under = undersample.fit_resample(X_scaled, y)

# Print distribusi kelas setelah undersampling
print(f'Distribusi kelas sebelum undersampling: {Counter(y)}')
print(f'Distribusi kelas setelah undersampling: {Counter(y_under)}')

# Optional: Tampilkan beberapa baris data hasil resampling
print(pd.DataFrame(X_under, columns=X.columns).head())


Distribusi kelas sebelum undersampling: Counter({0: 278843, 1: 24396})
Distribusi kelas setelah undersampling: Counter({0: 30495, 1: 24396})
   NAME_CONTRACT_TYPE  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  \
0            0.324189     1.390300      1.397564         0.665112   
1            0.324189    -0.719269     -0.715531        -1.503506   
2            0.324189     1.390300      1.397564         0.665112   
3            0.324189    -0.719269     -0.715531         0.665112   
4            0.324189    -0.719269      1.397564         0.665112   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0     -0.589199          1.222404    0.074070     0.675279   
1     -0.589199         -0.926918   -1.122195    -0.619653   
2      2.528457         -2.250760    0.492988    -0.822423   
3     -0.589199          0.118904   -0.296762     0.741823   
4     -0.589199         -1.359533   -1.713667    -1.103239   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  ...  \
0                    

# Test Model

## Model Initialization

In [ ]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, make_scorer, fbeta_score
import pandas as pd

# Define function to train model and calculate metrics
def evaluate_model(model, X_train, X_test, y_train, y_test):
    model.fit(X_train, y_train)

    # Predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    # Metrics calculation
    metrics = {
        "Accuracy (Train Set)": accuracy_score(y_train, y_train_pred),
        "Accuracy (Test Set)": accuracy_score(y_test, y_test_pred),
        "Precision (Test Set)": precision_score(y_test, y_test_pred),
        "Recall (Test Set)": recall_score(y_test, y_test_pred),
        "F1-Score (Test Set)": f1_score(y_test, y_test_pred),
        "F2-Score (Test Set)": fbeta_score(y_test, y_test_pred, beta=2),
    }

    # Calculate ROC AUC if the model supports predict_proba
    try:
        metrics["ROC AUC (Train Set)"] = roc_auc_score(y_train, model.predict_proba(X_train)[:, 1])
        metrics["ROC AUC (Test Set)"] = roc_auc_score(y_test, model.predict_proba(X_test)[:, 1])
    except AttributeError:
        metrics["ROC AUC (Train Set)"] = None
        metrics["ROC AUC (Test Set)"] = None

    # Cross-validation recall score
    metrics["Recall (Cross-Validation Train)"] = cross_val_score(model, X_train, y_train, cv=5, scoring=make_scorer(recall_score)).mean()
    metrics["Recall (Cross-Validation Test)"] = cross_val_score(model, X_test, y_test, cv=5, scoring=make_scorer(recall_score)).mean()

    return metrics

# Splitting the data
X_train_orig, X_test_orig, y_train_orig, y_test_orig = train_test_split(X_scaled, y, test_size=0.3, random_state=42)
X_train_over, X_test_over, y_train_over, y_test_over = train_test_split(X_smote, y_smote, test_size=0.3, random_state=42)
X_train_under, X_test_under, y_train_under, y_test_under = train_test_split(X_under, y_under, test_size=0.3, random_state=42)



## 1. Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression

# Initialize the Logistic Regression model (commonly used for classification tasks)
model = LogisticRegression()

# Evaluate the model using original data
metrics_orig = evaluate_model(model, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data:")
for metric, value in metrics_orig.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Original Data:
Accuracy (Train Set): 0.9194
Accuracy (Test Set): 0.9200
Precision (Test Set): 0.5000
Recall (Test Set): 0.0001
F1-Score (Test Set): 0.0003
F2-Score (Test Set): 0.0002
ROC AUC (Train Set): 0.6770
ROC AUC (Test Set): 0.6799
Recall (Cross-Validation Train): 0.0002
Recall (Cross-Validation Test): 0.0001


### Hyperparameter Tuning Logistic Regression <br>
Original

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

# Define the parameter grid
param_grid = {
    'C': [0.01, 0.1, 1, 10, 100],  # Regularization parameter
    'solver': ['liblinear', 'lbfgs'],  # Different solvers
    'max_iter': [100, 200, 500]  # Maximum number of iterations
}

# Initialize the Logistic Regression model
log_reg = LogisticRegression(random_state=42)

# Set up GridSearchCV
grid_search = GridSearchCV(estimator=log_reg, param_grid=param_grid,
                           scoring='accuracy', cv=5, n_jobs=-1, verbose=1)

# Perform the grid search on the training data
grid_search.fit(X_train_orig, y_train_orig)

# Retrieve the best model and parameters
best_log_reg = grid_search.best_estimator_
best_params = grid_search.best_params_

# Print the best parameters and the corresponding accuracy
print("Best Parameters for Logistic Regression:", best_params)
print("Best Cross-Validation Accuracy:", grid_search.best_score_)


Fitting 5 folds for each of 30 candidates, totalling 150 fits
Best Parameters for Logistic Regression: {'C': 0.1, 'max_iter': 100, 'solver': 'liblinear'}
Best Cross-Validation Accuracy: 0.9193657044585297


In [ ]:
from sklearn.linear_model import LogisticRegression

# Initialize the Logistic Regression model with the best parameters
model = LogisticRegression(C=0.1, max_iter=100, solver='liblinear')

# Evaluate the model using original data
metrics_orig = evaluate_model(model, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics

print("Evaluation Metrics on Original Data:")
for metric, value in metrics_orig.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Original Data:
Accuracy (Train Set): 0.9194
Accuracy (Test Set): 0.9200
Precision (Test Set): 0.5000
Recall (Test Set): 0.0001
F1-Score (Test Set): 0.0003
F2-Score (Test Set): 0.0002
ROC AUC (Train Set): 0.6776
ROC AUC (Test Set): 0.6805
Recall (Cross-Validation Train): 0.0002
Recall (Cross-Validation Test): 0.0001


### 1a. Oversampling

In [ ]:
# Evaluate the model using oversampled data
metrics_over = evaluate_model(model, X_train_over, X_test_over, y_train_over, y_test_over)

# Display the metrics
print("Evaluation Metrics on Oversampled Data:")
for metric, value in metrics_over.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Oversampled Data:
Accuracy (Train Set): 0.6935
Accuracy (Test Set): 0.6922
Precision (Test Set): 0.5775
Recall (Test Set): 0.2894
F1-Score (Test Set): 0.3856
F2-Score (Test Set): 0.3215
ROC AUC (Train Set): 0.7071
ROC AUC (Test Set): 0.7042
Recall (Cross-Validation Train): 0.2908
Recall (Cross-Validation Test): 0.2874


### 1b. Undersampling

In [ ]:
from sklearn.linear_model import LogisticRegression

# Initialize the Logistic Regression model with chosen parameters for undersampling evaluation
model_lr = LogisticRegression(solver='liblinear', random_state=42)

# Evaluate the Logistic Regression model using undersampled data
metrics_lr_under = evaluate_model(model_lr, X_train_under, X_test_under, y_train_under, y_test_under)

# Display the metrics for the undersampled data
print("Evaluation Metrics on Undersampled Data with Logistic Regression:")
for metric, value in metrics_lr_under.items():
    print(f"{metric}: {value:.4f}")



Evaluation Metrics on Undersampled Data with Logistic Regression:
Accuracy (Train Set): 0.6405
Accuracy (Test Set): 0.6391
Precision (Test Set): 0.6157
Recall (Test Set): 0.5093
F1-Score (Test Set): 0.5575
F2-Score (Test Set): 0.5275
ROC AUC (Train Set): 0.6843
ROC AUC (Test Set): 0.6762
Recall (Cross-Validation Train): 0.5005
Recall (Cross-Validation Test): 0.5008


## 2. kNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

# Initialize the kNN model with chosen parameters
model_knn = KNeighborsClassifier(n_neighbors=5, weights='distance', p=2)

# Evaluate the kNN model using original data
metrics_knn_orig = evaluate_model(model_knn, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with kNN:")
for metric, value in metrics_knn_orig.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Original Data with kNN:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.9146
Precision (Test Set): 0.1484
Recall (Test Set): 0.0142
F1-Score (Test Set): 0.0258
F2-Score (Test Set): 0.0173
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.5435
Recall (Cross-Validation Train): 0.0147
Recall (Cross-Validation Test): 0.0148


### Hyperparameter kNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

# Define parameter grid for kNN
param_grid_knn = {
    'n_neighbors': [3, 5, 7, 9, 11],
    'weights': ['uniform', 'distance'],
    'p': [1, 2]  # 1 for Manhattan distance, 2 for Euclidean distance
}

# Initialize kNN model
model_knn = KNeighborsClassifier()

# Set up GridSearchCV
grid_search_knn = GridSearchCV(estimator=model_knn, param_grid=param_grid_knn, cv=5, scoring='f1', n_jobs=-1, verbose=1)

# Fit the grid search to find the best parameters
grid_search_knn.fit(X_train_orig, y_train_orig)

# Get the best model with the best parameters
best_knn_model = grid_search_knn.best_estimator_
print("Best Parameters for kNN:", grid_search_knn.best_params_)

# Evaluate the best model using the original data
metrics_knn_best = evaluate_model(best_knn_model, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with Best kNN Model:")
for metric, value in metrics_knn_best.items():
    print(f"{metric}: {value:.4f}")

Fitting 5 folds for each of 20 candidates, totalling 100 fits
Best Parameters for kNN: {'n_neighbors': 3, 'p': 1, 'weights': 'distance'}
Evaluation Metrics on Original Data with Best kNN Model:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.9033
Precision (Test Set): 0.1213
Recall (Test Set): 0.0335
F1-Score (Test Set): 0.0525
F2-Score (Test Set): 0.0392
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.5334
Recall (Cross-Validation Train): 0.0386
Recall (Cross-Validation Test): 0.0393


### 2a. Oversampling

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

# Initialize the kNN model with chosen parameters for oversampling evaluation
model_knn = KNeighborsClassifier(n_neighbors=5, weights='distance', p=2)

# Evaluate the kNN model using oversampled data
metrics_knn_over = evaluate_model(model_knn, X_train_over, X_test_over, y_train_over, y_test_over)

# Display the metrics
print("Evaluation Metrics on Oversampled Data with kNN:")
for metric, value in metrics_knn_over.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Oversampled Data with kNN:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.7757
Precision (Test Set): 0.5996
Recall (Test Set): 0.9870
F1-Score (Test Set): 0.7460
F2-Score (Test Set): 0.8741
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.9445
Recall (Cross-Validation Train): 0.9755
Recall (Cross-Validation Test): 0.8955


### 2b. Undersampling

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

# Initialize the kNN model with chosen parameters for undersampling evaluation
model_knn = KNeighborsClassifier(n_neighbors=5, weights='distance', p=2)

# Evaluate the kNN model using undersampled data
metrics_knn_under = evaluate_model(model_knn, X_train_under, X_test_under, y_train_under, y_test_under)

# Display the metrics
print("Evaluation Metrics on Undersampled Data with kNN:")
for metric, value in metrics_knn_under.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Undersampled Data with kNN:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.5556
Precision (Test Set): 0.5023
Recall (Test Set): 0.4721
F1-Score (Test Set): 0.4867
F2-Score (Test Set): 0.4779
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.5640
Recall (Cross-Validation Train): 0.4808
Recall (Cross-Validation Test): 0.4737


## 3. Decision Tree

In [ ]:
from sklearn.tree import DecisionTreeClassifier

# Initialize the Decision Tree model with chosen parameters (you can adjust max_depth, criterion, etc. if needed)
model_tree = DecisionTreeClassifier()

# Evaluate the Decision Tree model using original data
metrics_tree_orig = evaluate_model(model_tree, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with Decision Tree:")
for metric, value in metrics_tree_orig.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Original Data with Decision Tree:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.8471
Precision (Test Set): 0.1133
Recall (Test Set): 0.1336
F1-Score (Test Set): 0.1226
F2-Score (Test Set): 0.1290
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.5214
Recall (Cross-Validation Train): 0.1261
Recall (Cross-Validation Test): 0.1399


#### Hyperparameter Decision Tree

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

# Define parameter grid
param_grid = {
    'max_depth': [None, 10, 20, 30, 40, 50],
    'min_samples_split': [2, 5, 10, 20],
    'min_samples_leaf': [1, 2, 5, 10],
    'criterion': ['gini', 'entropy']
}

# Initialize Decision Tree model
model_tree = DecisionTreeClassifier(random_state=42)

# Set up GridSearchCV
grid_search = GridSearchCV(estimator=model_tree, param_grid=param_grid, cv=5, scoring='f1', n_jobs=-1, verbose=1)

# Fit the grid search to find the best parameters
grid_search.fit(X_train_orig, y_train_orig)

# Get the best model with the best parameters
best_tree_model = grid_search.best_estimator_
print("Best Parameters for Decision Tree:", grid_search.best_params_)

# Evaluate the best model using the original data
metrics_tree_best = evaluate_model(best_tree_model, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with Best Decision Tree Model:")
for metric, value in metrics_tree_best.items():
    print(f"{metric}: {value:.4f}")


Fitting 5 folds for each of 192 candidates, totalling 960 fits
Best Parameters for Decision Tree: {'criterion': 'entropy', 'max_depth': 50, 'min_samples_leaf': 1, 'min_samples_split': 2}
Evaluation Metrics on Original Data with Best Decision Tree Model:
Accuracy (Train Set): 0.9998
Accuracy (Test Set): 0.8535
Precision (Test Set): 0.1210
Recall (Test Set): 0.1327
F1-Score (Test Set): 0.1266
F2-Score (Test Set): 0.1302
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.5242
Recall (Cross-Validation Train): 0.1304
Recall (Cross-Validation Test): 0.1340


### 3a. Oversampling

In [ ]:
from sklearn.tree import DecisionTreeClassifier

# Define the Decision Tree model
model_tree = DecisionTreeClassifier()

# Fit the model to the oversampled training data
model_tree.fit(X_train_over, y_train_over)

# Evaluate the model
metrics_tree_over = evaluate_model(model_tree, X_train_over, X_test_over, y_train_over, y_test_over)

# Display the metrics
print("Evaluation Metrics on Oversampled Data with Decision Tree:")
for metric, value in metrics_tree_over.items():
    print(f"{metric}: {value:.4f}")



Evaluation Metrics on Oversampled Data with Decision Tree:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.8761
Precision (Test Set): 0.8060
Recall (Test Set): 0.8281
F1-Score (Test Set): 0.8169
F2-Score (Test Set): 0.8236
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.8641
Recall (Cross-Validation Train): 0.8250
Recall (Cross-Validation Test): 0.7995


### 3b. Undersampling

In [ ]:
from sklearn.tree import DecisionTreeClassifier

# Define the Decision Tree model
model_tree = DecisionTreeClassifier()

# Fit the model to the undersampled training data
model_tree.fit(X_train_under, y_train_under)

# Evaluate the model
metrics_tree_under = evaluate_model(model_tree, X_train_under, X_test_under, y_train_under, y_test_under)

# Display the metrics
print("Evaluation Metrics on Undersampled Data with Decision Tree:")
for metric, value in metrics_tree_under.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Undersampled Data with Decision Tree:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.5533
Precision (Test Set): 0.4996
Recall (Test Set): 0.5023
F1-Score (Test Set): 0.5009
F2-Score (Test Set): 0.5018
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.5484
Recall (Cross-Validation Train): 0.4945
Recall (Cross-Validation Test): 0.5184


## 4. Bagging: Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Initialize the Random Forest model with chosen parameters (you can adjust n_estimators, max_depth, etc.)
model_rf = RandomForestClassifier(random_state=42)

# Evaluate the Random Forest model using original data
metrics_rf_orig = evaluate_model(model_rf, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with Random Forest:")
for metric, value in metrics_rf_orig.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Original Data with Random Forest:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.9200
Precision (Test Set): 1.0000
Recall (Test Set): 0.0001
F1-Score (Test Set): 0.0003
F2-Score (Test Set): 0.0002
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.6544
Recall (Cross-Validation Train): 0.0002
Recall (Cross-Validation Test): 0.0000


### Hyperparameter Bagging: Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

# Define parameter grid for Random Forest
param_grid_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20, 30, 40],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'criterion': ['gini', 'entropy']
}

# Initialize Random Forest model
model_rf = RandomForestClassifier(random_state=42)

# Set up GridSearchCV
grid_search_rf = GridSearchCV(estimator=model_rf, param_grid=param_grid_rf, cv=5, scoring='f1', n_jobs=-1, verbose=1)

# Fit the grid search to find the best parameters
grid_search_rf.fit(X_train_orig, y_train_orig)

# Get the best model with the best parameters
best_rf_model = grid_search_rf.best_estimator_
print("Best Parameters for Random Forest:", grid_search_rf.best_params_)

# Evaluate the best model using the original data
metrics_rf_best = evaluate_model(best_rf_model, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with Best Random Forest Model:")
for metric, value in metrics_rf_best.items():
    print(f"{metric}: {value:.4f}")


Fitting 5 folds for each of 270 candidates, totalling 1350 fits
Best Parameters for Random Forest: {'criterion': 'gini', 'max_depth': 40, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100}
Evaluation Metrics on Original Data with Best Random Forest Model:
Accuracy (Train Set): 0.9998
Accuracy (Test Set): 0.9200
Precision (Test Set): 0.0000
Recall (Test Set): 0.0000
F1-Score (Test Set): 0.0000
F2-Score (Test Set): 0.0000
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.6564
Recall (Cross-Validation Train): 0.0003
Recall (Cross-Validation Test): 0.0000


### 4a. Oversampling

In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Define the Random Forest model
model_rf = RandomForestClassifier()

# Fit the model to the oversampled training data
model_rf.fit(X_train_over, y_train_over)

# Evaluate the model
metrics_rf_over = evaluate_model(model_rf, X_train_over, X_test_over, y_train_over, y_test_over)

# Display the metrics
print("Evaluation Metrics on Oversampled Data with Random Forest:")
for metric, value in metrics_rf_over.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Oversampled Data with Random Forest:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.9370
Precision (Test Set): 0.9996
Recall (Test Set): 0.8117
F1-Score (Test Set): 0.8959
F2-Score (Test Set): 0.8434
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.9655
Recall (Cross-Validation Train): 0.8076
Recall (Cross-Validation Test): 0.7778


### 4b. Undersampling

In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Define the Random Forest model
model_rf = RandomForestClassifier()

# Fit the model to the undersampled training data
model_rf.fit(X_train_under, y_train_under)

# Evaluate the model
metrics_rf_under = evaluate_model(model_rf, X_train_under, X_test_under, y_train_under, y_test_under)

# Display the metrics
print("Evaluation Metrics on Undersampled Data with Random Forest:")
for metric, value in metrics_rf_under.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Undersampled Data with Random Forest:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.6364
Precision (Test Set): 0.6254
Recall (Test Set): 0.4622
F1-Score (Test Set): 0.5315
F2-Score (Test Set): 0.4876
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.6717
Recall (Cross-Validation Train): 0.4554
Recall (Cross-Validation Test): 0.4522


## 5. Boosting: Adaboost

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

# Initialize the AdaBoost model with chosen parameters (you can adjust n_estimators, learning_rate, etc.)
model_ab = AdaBoostClassifier(n_estimators=50, learning_rate=1.0, random_state=42)

# Evaluate the AdaBoost model using original data
metrics_ab_orig = evaluate_model(model_ab, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with AdaBoost:")
for metric, value in metrics_ab_orig.items():
    print(f"{metric}: {value:.4f}")

Evaluation Metrics on Original Data with AdaBoost:
Accuracy (Train Set): 0.9194
Accuracy (Test Set): 0.9200
Precision (Test Set): 1.0000
Recall (Test Set): 0.0001
F1-Score (Test Set): 0.0003
F2-Score (Test Set): 0.0002
ROC AUC (Train Set): 0.6910
ROC AUC (Test Set): 0.6933
Recall (Cross-Validation Train): 0.0002
Recall (Cross-Validation Test): 0.0001


### Hyperparameter Boosting: Adaboost

In [ ]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier

# Define parameter grid for AdaBoost
param_grid_ab = {
    'n_estimators': [50, 100, 150],
    'learning_rate': [0.5, 1.0, 1.5],
    'estimator': [DecisionTreeClassifier(max_depth=1), DecisionTreeClassifier(max_depth=2)]
}

# Initialize AdaBoost model
model_ab = AdaBoostClassifier(random_state=42)

# Set up GridSearchCV
grid_search_ab = GridSearchCV(estimator=model_ab, param_grid=param_grid_ab, cv=5, scoring='f1', n_jobs=-1, verbose=1)

# Fit the grid search to find the best parameters
grid_search_ab.fit(X_train_orig, y_train_orig)

# Get the best model with the best parameters
best_ab_model = grid_search_ab.best_estimator_
print("Best Parameters for AdaBoost:", grid_search_ab.best_params_)

# Evaluate the best model using the original data
metrics_ab_best = evaluate_model(best_ab_model, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with Best AdaBoost Model:")
for metric, value in metrics_ab_best.items():
    print(f"{metric}: {value:.4f}")

Fitting 5 folds for each of 18 candidates, totalling 90 fits
Best Parameters for AdaBoost: {'estimator': DecisionTreeClassifier(max_depth=2), 'learning_rate': 1.5, 'n_estimators': 150}
Evaluation Metrics on Original Data with Best AdaBoost Model:
Accuracy (Train Set): 0.9189
Accuracy (Test Set): 0.9190
Precision (Test Set): 0.3168
Recall (Test Set): 0.0114
F1-Score (Test Set): 0.0220
F2-Score (Test Set): 0.0141
ROC AUC (Train Set): 0.7329
ROC AUC (Test Set): 0.6969
Recall (Cross-Validation Train): 0.0134
Recall (Cross-Validation Test): 0.0242


### 5a. Oversampling

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

# Define the AdaBoost model
model_ab = AdaBoostClassifier()

# Fit the model to the oversampled training data
model_ab.fit(X_train_over, y_train_over)

# Evaluate the model
metrics_ab_over = evaluate_model(model_ab, X_train_over, X_test_over, y_train_over, y_test_over)

# Display the metrics
print("Evaluation Metrics on Oversampled Data with AdaBoost:")
for metric, value in metrics_ab_over.items():
    print(f"{metric}: {value:.4f}")

Evaluation Metrics on Oversampled Data with AdaBoost:
Accuracy (Train Set): 0.8974
Accuracy (Test Set): 0.8967
Precision (Test Set): 0.9669
Recall (Test Set): 0.7150
F1-Score (Test Set): 0.8221
F2-Score (Test Set): 0.7543
ROC AUC (Train Set): 0.9214
ROC AUC (Test Set): 0.9199
Recall (Cross-Validation Train): 0.7144
Recall (Cross-Validation Test): 0.7168


### 5b Undersampling

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

# Define the AdaBoost model
model_ab = AdaBoostClassifier()

# Fit the model to the undersampled training data
model_ab.fit(X_train_under, y_train_under)

# Evaluate the model
metrics_ab_under = evaluate_model(model_ab, X_train_under, X_test_under, y_train_under, y_test_under)

# Display the metrics
print("Evaluation Metrics on Undersampled Data with AdaBoost:")
for metric, value in metrics_ab_under.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Undersampled Data with AdaBoost:
Accuracy (Train Set): 0.6455
Accuracy (Test Set): 0.6449
Precision (Test Set): 0.6266
Recall (Test Set): 0.5056
F1-Score (Test Set): 0.5596
F2-Score (Test Set): 0.5259
ROC AUC (Train Set): 0.6916
ROC AUC (Test Set): 0.6876
Recall (Cross-Validation Train): 0.5118
Recall (Cross-Validation Test): 0.5159


## 6. Boosting: XGBoost

In [ ]:
pip install xgboost

Note: you may need to restart the kernel to use updated packages.


In [ ]:
from xgboost import XGBClassifier
import pickle

# Initialize the XGBoost model with chosen parameters (you can adjust n_estimators, learning_rate, max_depth, etc.)
model_xgb = XGBClassifier(n_estimators=100, learning_rate=0.1, max_depth=6, random_state=42, use_label_encoder=False, eval_metric="logloss")

# Evaluate the XGBoost model using original data
metrics_xgb_orig = evaluate_model(model_xgb, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with XGBoost:")
for metric, value in metrics_xgb_orig.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Original Data with XGBoost:
Accuracy (Train Set): 0.9195
Accuracy (Test Set): 0.9200
Precision (Test Set): 0.5385
Recall (Test Set): 0.0010
F1-Score (Test Set): 0.0019
F2-Score (Test Set): 0.0012
ROC AUC (Train Set): 0.7726
ROC AUC (Test Set): 0.7109
Recall (Cross-Validation Train): 0.0006
Recall (Cross-Validation Test): 0.0025


In [ ]:
import pickle
model_name = "xgb_model.pkl"
with open(model_name, "wb") as file:
    pickle.dump(model_xgb, file)

### Hyperparameter Boosting: XGBoost

In [ ]:
from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV

# Define parameter grid for XGBoost
param_grid_xgb = {
    'n_estimators': [100, 150, 200],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 5, 6, 7],
    'subsample': [0.6, 0.8, 1.0],
    'colsample_bytree': [0.6, 0.8, 1.0]
}

# Initialize XGBoost model
model_xgb = XGBClassifier(use_label_encoder=False, eval_metric="logloss", random_state=42)

# Set up GridSearchCV
grid_search_xgb = GridSearchCV(estimator=model_xgb, param_grid=param_grid_xgb, cv=5, scoring='f1', n_jobs=-1, verbose=1)

# Fit the grid search to find the best parameters
grid_search_xgb.fit(X_train_orig, y_train_orig)

# Get the best model with the best parameters
best_xgb_model = grid_search_xgb.best_estimator_
print("Best Parameters for XGBoost:", grid_search_xgb.best_params_)

# Evaluate the best model using the original data
metrics_xgb_best = evaluate_model(best_xgb_model, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with Best XGBoost Model:")
for metric, value in metrics_xgb_best.items():
    print(f"{metric}: {value:.4f}")


Fitting 5 folds for each of 324 candidates, totalling 1620 fits
Best Parameters for XGBoost: {'colsample_bytree': 0.8, 'learning_rate': 0.2, 'max_depth': 7, 'n_estimators': 200, 'subsample': 0.6}
Evaluation Metrics on Original Data with Best XGBoost Model:
Accuracy (Train Set): 0.9251
Accuracy (Test Set): 0.9192
Precision (Test Set): 0.3590
Recall (Test Set): 0.0135
F1-Score (Test Set): 0.0260
F2-Score (Test Set): 0.0167
ROC AUC (Train Set): 0.9016
ROC AUC (Test Set): 0.6893
Recall (Cross-Validation Train): 0.0138
Recall (Cross-Validation Test): 0.0197


### 6a. Oversampling

In [ ]:
import xgboost as xgb

# Define the XGBoost model
model_xgb = xgb.XGBClassifier(device='cuda') #pake runtime T4 GPU

# Fit the model to the oversampled training data
model_xgb.fit(X_train_over, y_train_over)

# Evaluate the model
metrics_xgb_over = evaluate_model(model_xgb, X_train_over, X_test_over, y_train_over, y_test_over)

# Display the metrics
print("Evaluation Metrics on Oversampled Data with XGBoost:")
for metric, value in metrics_xgb_over.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Oversampled Data with XGBoost:
Accuracy (Train Set): 0.9406
Accuracy (Test Set): 0.9387
Precision (Test Set): 0.9985
Recall (Test Set): 0.8176
F1-Score (Test Set): 0.8990
F2-Score (Test Set): 0.8483
ROC AUC (Train Set): 0.9646
ROC AUC (Test Set): 0.9469
Recall (Cross-Validation Train): 0.8197
Recall (Cross-Validation Test): 0.8165


In [ ]:
import pickle
model_name = "xgb_model.pkl"
with open(model_name, "wb") as file:
    pickle.dump(model_xgb, file)

### Hyperparameter (Oversampling)

In [ ]:
from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV

# Initialize XGBoost model
model_xgb = XGBClassifier(use_label_encoder=False, eval_metric='logloss', device='cuda')

# Define parameter grid for hyperparameter tuning
param_grid_xgb = {
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 5, 7],
    'n_estimators': [100, 200, 300],
    'subsample': [0.6, 0.8, 1.0],
    'colsample_bytree': [0.6, 0.8, 1.0]
}

# Set up GridSearchCV for hyperparameter tuning using the oversampled training data
grid_search_xgb = GridSearchCV(estimator=model_xgb, param_grid=param_grid_xgb, cv=5, scoring='f1', n_jobs=-1, verbose=1)
grid_search_xgb.fit(X_train_over, y_train_over)

# Get the best model with the best parameters
best_xgb_model = grid_search_xgb.best_estimator_
print("Best Parameters for XGBoost:", grid_search_xgb.best_params_)

# Evaluate the best model using the oversampled test data
metrics_xgb_over = evaluate_model(best_xgb_model, X_train_over, X_test_over, y_train_over, y_test_over)

# Display the metrics
print("Evaluation Metrics on Oversampled Data with Best XGBoost Model:")
for metric, value in metrics_xgb_over.items():
    print(f"{metric}: {value:.4f}")


Fitting 5 folds for each of 243 candidates, totalling 1215 fits
Best Parameters for XGBoost: {'colsample_bytree': 0.6, 'learning_rate': 0.2, 'max_depth': 7, 'n_estimators': 300, 'subsample': 1.0}
Evaluation Metrics on Oversampled Data with Best XGBoost Model:
Accuracy (Train Set): 0.9460
Accuracy (Test Set): 0.9397
Precision (Test Set): 0.9966
Recall (Test Set): 0.8220
F1-Score (Test Set): 0.9009
F2-Score (Test Set): 0.8519
ROC AUC (Train Set): 0.9861
ROC AUC (Test Set): 0.9470
Recall (Cross-Validation Train): 0.8239
Recall (Cross-Validation Test): 0.8208


### 6b. Undersampling

In [ ]:
import xgboost as xgb

# Define the XGBoost model
model_xgb = xgb.XGBClassifier()

# Fit the model to the undersampled training data
model_xgb.fit(X_train_under, y_train_under)

# Evaluate the model
metrics_xgb_under = evaluate_model(model_xgb, X_train_under, X_test_under, y_train_under, y_test_under)

# Display the metrics
print("Evaluation Metrics on Undersampled Data with XGBoost:")
for metric, value in metrics_xgb_under.items():
    print(f"{metric}: {value:.4f}")



Evaluation Metrics on Undersampled Data with XGBoost:
Accuracy (Train Set): 0.7960
Accuracy (Test Set): 0.6431
Precision (Test Set): 0.6150
Recall (Test Set): 0.5361
F1-Score (Test Set): 0.5728
F2-Score (Test Set): 0.5502
ROC AUC (Train Set): 0.8808
ROC AUC (Test Set): 0.6898
Recall (Cross-Validation Train): 0.5343
Recall (Cross-Validation Test): 0.5348


#### Hyperparameter (Undersampling)

In [ ]:
from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV

# Initialize XGBoost model
model_xgb = XGBClassifier(use_label_encoder=False, eval_metric='logloss')

# Define parameter grid for hyperparameter tuning
param_grid_xgb = {
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 5, 7],
    'n_estimators': [100, 200, 300],
    'subsample': [0.6, 0.8, 1.0],
    'colsample_bytree': [0.6, 0.8, 1.0]
}

# Set up GridSearchCV for hyperparameter tuning using the undersampled training data
grid_search_xgb = GridSearchCV(estimator=model_xgb, param_grid=param_grid_xgb, cv=5, scoring='f1', n_jobs=-1, verbose=1)
grid_search_xgb.fit(X_train_under, y_train_under)

# Get the best model with the best parameters
best_xgb_model = grid_search_xgb.best_estimator_
print("Best Parameters for XGBoost:", grid_search_xgb.best_params_)

# Evaluate the best model using the undersampled test data
metrics_xgb_under = evaluate_model(best_xgb_model, X_train_under, X_test_under, y_train_under, y_test_under)

# Display the metrics
print("Evaluation Metrics on Undersampled Data with Best XGBoost Model:")
for metric, value in metrics_xgb_under.items():
    print(f"{metric}: {value:.4f}")


Fitting 5 folds for each of 243 candidates, totalling 1215 fits
Best Parameters for XGBoost: {'colsample_bytree': 0.6, 'learning_rate': 0.2, 'max_depth': 3, 'n_estimators': 300, 'subsample': 0.8}
Evaluation Metrics on Undersampled Data with Best XGBoost Model:
Accuracy (Train Set): 0.7004
Accuracy (Test Set): 0.6578
Precision (Test Set): 0.6385
Recall (Test Set): 0.5378
F1-Score (Test Set): 0.5839
F2-Score (Test Set): 0.5553
ROC AUC (Train Set): 0.7697
ROC AUC (Test Set): 0.7112
Recall (Cross-Validation Train): 0.5398
Recall (Cross-Validation Test): 0.5412


## 7. Stacking

In [ ]:
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

# Define base estimators
estimators = [
    ('rf', RandomForestClassifier(n_estimators=100, random_state=42)),
    ('knn', KNeighborsClassifier(n_neighbors=5, weights='distance')),
    ('log_reg', LogisticRegression())
]

# Initialize the Stacking model with Logistic Regression as the meta-learner
model_stacking = StackingClassifier(estimators=estimators, final_estimator=LogisticRegression(), cv=5)

# Evaluate the Stacking model using original data
metrics_stacking_orig = evaluate_model(model_stacking, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with Stacking:")
for metric, value in metrics_stacking_orig.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Original Data with Stacking:
Accuracy (Train Set): 0.9612
Accuracy (Test Set): 0.9196
Precision (Test Set): 0.3299
Recall (Test Set): 0.0044
F1-Score (Test Set): 0.0087
F2-Score (Test Set): 0.0055
ROC AUC (Train Set): 0.9998
ROC AUC (Test Set): 0.6847
Recall (Cross-Validation Train): 0.0066
Recall (Cross-Validation Test): 0.0036


### Hyperparameter Stacking

In [ ]:
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

# Define base estimators
estimators = [
    ('rf', RandomForestClassifier(random_state=42)),
    ('knn', KNeighborsClassifier()),
    ('log_reg', LogisticRegression())
]

# Initialize Stacking Classifier with Logistic Regression as the meta-learner
model_stacking = StackingClassifier(estimators=estimators, final_estimator=LogisticRegression(), cv=5)

# Define parameter grid for hyperparameter tuning
param_grid_stacking = {
    'final_estimator__C': [0.01, 0.1, 1, 10],
    'rf__max_depth': [None, 10, 20, 30],
    'knn__n_neighbors': [3, 5, 7, 9],
}

# Set up GridSearchCV for hyperparameter tuning using the original training data
grid_search_stacking = GridSearchCV(estimator=model_stacking, param_grid=param_grid_stacking, cv=5, scoring='f1', n_jobs=-1, verbose=1)
grid_search_stacking.fit(X_train_orig, y_train_orig)

# Get the best model with the best parameters
best_stacking_model = grid_search_stacking.best_estimator_
print("Best Parameters for Stacking Classifier:", grid_search_stacking.best_params_)

# Evaluate the best model using the original test data
metrics_stacking_orig = evaluate_model(best_stacking_model, X_train_orig, X_test_orig, y_train_orig, y_test_orig)

# Display the metrics
print("Evaluation Metrics on Original Data with Best Stacking Model:")
for metric, value in metrics_stacking_orig.items():
    print(f"{metric}: {value:.4f}")

Fitting 5 folds for each of 64 candidates, totalling 320 fits
Best Parameters for Stacking Classifier: {'final_estimator__C': 10, 'knn__n_neighbors': 5, 'rf__max_depth': None}
Evaluation Metrics on Original Data with Best Stacking Model:
Accuracy (Train Set): 0.9627
Accuracy (Test Set): 0.9196
Precision (Test Set): 0.3391
Recall (Test Set): 0.0054
F1-Score (Test Set): 0.0106
F2-Score (Test Set): 0.0067
ROC AUC (Train Set): 0.9998
ROC AUC (Test Set): 0.6848
Recall (Cross-Validation Train): 0.0068
Recall (Cross-Validation Test): 0.0040


### 7a. Oversampling

In [ ]:
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

# Define base estimators
estimators = [
    ('decision_tree', DecisionTreeClassifier()),
    ('random_forest', RandomForestClassifier())
]

# Define the Stacking model with Logistic Regression as the meta-learner
model_stacking = StackingClassifier(estimators=estimators, final_estimator=LogisticRegression())

# Fit the model to the oversampled training data
model_stacking.fit(X_train_over, y_train_over)

# Evaluate the model
metrics_stacking_over = evaluate_model(model_stacking, X_train_over, X_test_over, y_train_over, y_test_over)

# Display the metrics
print("Evaluation Metrics on Oversampled Data with Stacking:")
for metric, value in metrics_stacking_over.items():
    print(f"{metric}: {value:.4f}")



Evaluation Metrics on Oversampled Data with Stacking:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.9443
Precision (Test Set): 0.9817
Recall (Test Set): 0.8489
F1-Score (Test Set): 0.9104
F2-Score (Test Set): 0.8725
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.9658
Recall (Cross-Validation Train): 0.8470
Recall (Cross-Validation Test): 0.8297


### 7b. Undersampling

In [ ]:
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

# Define base estimators
estimators = [
    ('decision_tree', DecisionTreeClassifier()),
    ('random_forest', RandomForestClassifier())
]

# Define the Stacking model
model_stacking = StackingClassifier(estimators=estimators, final_estimator=LogisticRegression())

# Fit the model to the undersampled training data
model_stacking.fit(X_train_under, y_train_under)

# Evaluate the model
metrics_stacking_under = evaluate_model(model_stacking, X_train_under, X_test_under, y_train_under, y_test_under)

# Display the metrics
print("Evaluation Metrics on Undersampled Data with Stacking:")
for metric, value in metrics_stacking_under.items():
    print(f"{metric}: {value:.4f}")


Evaluation Metrics on Undersampled Data with Stacking:
Accuracy (Train Set): 1.0000
Accuracy (Test Set): 0.6325
Precision (Test Set): 0.6146
Recall (Test Set): 0.4735
F1-Score (Test Set): 0.5349
F2-Score (Test Set): 0.4963
ROC AUC (Train Set): 1.0000
ROC AUC (Test Set): 0.6706
Recall (Cross-Validation Train): 0.4724
Recall (Cross-Validation Test): 0.4649


# Tabel Hasil Pemodelan

## Table 1. All Model (Data Biasa)


| Metric               | Logistic Regression | kNN    | Decision Tree | Random Forest | AdaBoost | XGBoost | Stacking |
|----------------------|---------------------|--------|---------------|---------------|----------|---------|----------|
| Accuracy (Train)     | 0.9194              | 1.0000 | 1.0000        | 1.0000        | 0.9194   | 0.9195  | 0.9612   |
| Accuracy (Test)      | 0.9200              | 0.9146 | 0.8469        | 0.9200        | 0.9200   | 0.9200  | 0.9196   |
| Precision (Test)     | 0.5000              | 0.1484 | 0.1121        | 1.0000        | 1.0000   | 0.5385  | 0.3299   |
| Recall (Test)        | 0.0001              | 0.0142 | 0.1322        | 0.0001        | 0.0001   | 0.0010  | 0.0044   |
| F1-Score (Test)      | 0.0003              | 0.0258 | 0.1213        | 0.0003        | 0.0003   | 0.0019  | 0.0087   |
| F2-Score (Test)      | 0.0002              | 0.0173 | 0.1276        | 0.0002        | 0.0002   | 0.0012  | 0.0055   |
| ROC AUC (Train)      | 0.6770              | 1.0000 | 1.0000        | 1.0000        | 0.6910   | 0.7726  | 0.9998   |
| ROC AUC (Test)       | 0.6799              | 0.5435 | 0.5206        | 0.6544        | 0.6933   | 0.7109  | 0.6847   |
| Recall (CV Train)    | 0.0002              | 0.0147 | 0.1260        | 0.0002        | 0.0002   | 0.0006  | 0.0066   |
| Recall (CV Test)     | 0.0001              | 0.0148 | 0.1387        | 0.0000        | 0.0001   | 0.0025  | 0.0036   |


**Semua model** pada data biasa memiliki recall dan F2-Score yang **rendah**, mengindikasikan banyak FN yang berpotensi meningkatkan risiko gagal bayar. Oleh karena itu, pendekatan ini kurang efektif untuk mencapai goal proyek

## Table 2. All Model (Oversampling)


| Metric               | Logistic Regression | kNN    | Decision Tree | Random Forest | AdaBoost | XGBoost | Stacking |
|----------------------|---------------------|--------|---------------|---------------|----------|---------|----------|
| Accuracy (Train)     | 0.6935              | 1.0000 | 1.0000        | 1.0000        | 0.8974   | 0.9408  | 1.0000   |
| Accuracy (Test)      | 0.6922              | 0.7757 | 0.8469        | 0.9370        | 0.8967   | 0.9390  | 0.9443   |
| Precision (Test)     | 0.5775              | 0.5996 | 0.1121        | 0.9996        | 0.9669   | 0.9985  | 0.9817   |
| Recall (Test)        | 0.2894              | 0.9870 | 0.1322        | 0.8117        | 0.7150   | 0.8183  | 0.8489   |
| F1-Score (Test)      | 0.3856              | 0.7460 | 0.1213        | 0.8959        | 0.8221   | 0.8995  | 0.9104   |
| F2-Score (Test)      | 0.3215              | 0.8741 | 0.1276        | 0.8434        | 0.7543   | 0.8489  | 0.8725   |
| ROC AUC (Train)      | 0.7071              | 1.0000 | 1.0000        | 1.0000        | 0.9214   | 0.9647  | 1.0000   |
| ROC AUC (Test)       | 0.7042              | 0.9445 | 0.5206        | 0.9655        | 0.9199   | 0.9471  | 0.9658   |
| Recall (CV Train)    | 0.2908              | 0.9755 | 0.1260        | 0.8076        | 0.7144   | 0.8205  | 0.8470   |
| Recall (CV Test)     | 0.2874              | 0.8955 | 0.1387        | 0.7778        | 0.7168   | 0.8175  | 0.8297   |


**Stacking** lalu diperingkat kedua **XGBoost** adalah model terbaik dalam pendekatan ini dengan F2-Score yang tinggi, menunjukkan keseimbangan yang baik antara deteksi nasabah berisiko dan inklusi finansial. Model ini memungkinkan deteksi yang akurat terhadap nasabah gagal bayar, yang sejalan dengan goal untuk mengurangi risiko kredit macet.

## Table 3. All Model (Undersampling)


| Metric                        | Logistic Regression | kNN    | Decision Tree | Random Forest | AdaBoost | XGBoost | Stacking |
|-------------------------------|---------------------|--------|---------------|---------------|----------|---------|----------|
| Accuracy (Train Set)          | 0.6405             | 1.0000 | 1.0000        | 1.0000        | 0.6455   | 0.7960  | 1.0000   |
| Accuracy (Test Set)           | 0.6391             | 0.5556 | 0.5533        | 0.6364        | 0.6449   | 0.6431  | 0.6325   |
| Precision (Test Set)          | 0.6157             | 0.5023 | 0.4996        | 0.6254        | 0.6266   | 0.6150  | 0.6146   |
| Recall (Test Set)             | 0.5093             | 0.4721 | 0.5023        | 0.4622        | 0.5056   | 0.5361  | 0.4735   |
| F1-Score (Test Set)           | 0.5575             | 0.4867 | 0.5009        | 0.5315        | 0.5596   | 0.5728  | 0.5349   |
| F2-Score (Test Set)           | 0.5275             | 0.4779 | 0.5018        | 0.4876        | 0.5259   | 0.5502  | 0.4963   |
| ROC AUC (Train Set)           | 0.6843             | 1.0000 | 1.0000        | 1.0000        | 0.6916   | 0.8808  | 1.0000   |
| ROC AUC (Test Set)            | 0.6762             | 0.5640 | 0.5484        | 0.6717        | 0.6876   | 0.6898  | 0.6706   |
| Recall (Cross-Validation Train)| 0.5005            | 0.4808 | 0.4945        | 0.4554        | 0.5118   | 0.5343  | 0.4724   |
| Recall (Cross-Validation Test) | 0.5008            | 0.4737 | 0.5184        | 0.4522        | 0.5159   | 0.5348  | 0.4649   |


**XGBoost** menunjukkan recall yang cukup baik (0.5361) dibandingkan data biasa, menurunkan FN namun menghasilkan lebih banyak FP karena lebih sering memprediksi defaulter. F2-Score (0.5553) menunjukkan peningkatan sensitivitas, tetapi tidak sebaik oversampling.


## Table 4. All Model (Hyperparameter Data Biasa)




| Metric               | Logistic Regression | kNN    | Decision Tree | Random Forest | AdaBoost | XGBoost | Stacking |
|----------------------|---------------------|--------|---------------|---------------|----------|---------|----------|
| Accuracy (Train)     | 0.9194              | 1.0000 | 0.9998        | 0.9998        | 0.9189   | 0.9251  | 0.9627   |
| Accuracy (Test)      | 0.9200              | 0.9033 | 0.8535        | 0.9200        | 0.9190   | 0.9192  | 0.9196   |
| Precision (Test)     | 0.5000              | 0.1213 | 0.1210        | 0.0000        | 0.3168   | 0.3590  | 0.3391   |
| Recall (Test)        | 0.0001              | 0.0335 | 0.1327        | 0.0000        | 0.0114   | 0.0135  | 0.0054   |
| F1-Score (Test)      | 0.0003              | 0.0525 | 0.1266        | 0.0000        | 0.0220   | 0.0260  | 0.0106   |
| F2-Score (Test)      | 0.0002              | 0.0392 | 0.1302        | 0.0000        | 0.0141   | 0.0167  | 0.0067   |
| ROC AUC (Train)      | 0.6776              | 1.0000 | 1.0000        | 1.0000        | 0.7329   | 0.9016  | 0.9998   |
| ROC AUC (Test)       | 0.6805              | 0.5334 | 0.5242        | 0.6564        | 0.6969   | 0.6893  | 0.6848   |
| Recall (CV Train)    | 0.0002              | 0.0386 | 0.1304        | 0.0003        | 0.0134   | 0.0138  | 0.0068   |
| Recall (CV Test)     | 0.0001              | 0.0393 | 0.1340        | 0.0000        | 0.0242   | 0.0197  | 0.0040   |


Setelah tuning, **semua model** pada data biasa memiliki nilai **recall** yang **sangat rendah**, membuat mereka kurang ideal untuk mendeteksi nasabah berisiko gagal bayar. Model ini tidak mencapai performa yang diinginkan dalam hal recall dan lebih baik dievaluasi ulang dengan pendekatan yang berbeda, seperti oversampling atau undersampling.

## Table 5. Best Model of Undersampling (Before & After Hyperparameter)


| Metric                          | XGBoost Before Tuning | XGBoost After Tuning |
|---------------------------------|-----------------------|-----------------------|
| Accuracy (Train Set)            | 0.9408                | 0.7004                |
| Accuracy (Test Set)             | 0.9390                | 0.6578                |
| Precision (Test Set)            | 0.9985                | 0.6385                |
| Recall (Test Set)               | 0.8183                | 0.5378                |
| F1-Score (Test Set)             | 0.8995                | 0.5839                |
| F2-Score (Test Set)             | 0.8489                | 0.5553                |
| ROC AUC (Train Set)             | 0.9647                | 0.7697                |
| ROC AUC (Test Set)              | 0.9471                | 0.7112                |
| Recall (Cross-Validation Train) | 0.8205                | 0.5398                |
| Recall (Cross-Validation Test)  | 0.8175                | 0.5412                |


Berdasarkan goal utama proyek yaitu secara agresif mengurangi risiko gagal bayar dengan mendeteksi sebanyak mungkin nasabah berisiko, maka **model sebelum tuning** lebih sesuai karena memiliki recall dan F2-Score yang lebih tinggi, meskipun ada risiko overfitting. Model ini akan lebih sensitif terhadap nasabah berisiko, mengurangi jumlah defaulter yang tidak terdeteksi (FN), meskipun ada peningkatan risiko FP dan overfitting.<br>

Namun, jika stabilitas model dan kemampuan generalisasi untuk data baru menjadi pertimbangan penting, terutama untuk implementasi jangka panjang, maka model sesudah tuning lebih direkomendasikan. Meskipun recall menurun, model ini lebih seimbang dan lebih cocok untuk kondisi data yang mungkin berubah di masa depan.

Pada data original dan Undersampling, hasil tuning pada model-model ini justru menurunkan performa utama yang dibutuhkan untuk mencapai tujuan proyek. Penurunan recall dan ROC AUC menunjukkan bahwa model menjadi kurang efektif dalam mendeteksi nasabah berisiko, yang dapat berakibat pada meningkatnya risiko kredit macet dan menurunnya inklusi finansial. **Model dengan parameter default dan menggunakan oversampling**, terutama pada **Stacking** dan XGBoost, memberikan hasil yang lebih baik untuk tujuan proyek, karena mereka mempertahankan nilai recall dan ROC AUC yang tinggi tanpa
menyebabkan overfitting.

## Table 6. The Best Model: Oversampling Without Tuning


| Metric               | XGBoost | Stacking |
|----------------------|---------|----------|
| Accuracy (Train)     | 0.9408  | 1.0000   |
| Accuracy (Test)      | 0.9390  | 0.9443   |
| Precision (Test)     | 0.9985  | 0.9817   |
| Recall (Test)        | 0.8183  | 0.8489   |
| F1-Score (Test)      | 0.8995  | 0.9104   |
| F2-Score (Test)      | 0.8489  | 0.8725   |
| ROC AUC (Train)      | 0.9647  | 1.0000   |
| ROC AUC (Test)       | 0.9471  | 0.9658   |
| Recall (CV Train)    | 0.8205  | 0.8470   |
| Recall (CV Test)     | 0.8175  | 0.8297   |


**Kesimpulan:**

1. **Modeling**: Implementasi menggunakan **XGBoost** dan **Stacking** untuk menangani data oversampling dengan algoritma yang lebih kompleks dari kelas.
  
2. **Model Evaluation**: **XGBoost** unggul dalam **ROC AUC** (0.9471) dan **Recall**, sesuai dengan tujuan proyek untuk mendeteksi risiko default. **Stacking** menunjukkan performa sedikit lebih baik pada beberapa metrik (seperti **F1-Score**), tetapi lebih kompleks dan memakan waktu.

3. **Best Fit & Validasi**: **XGBoost** terbukti **best-fit** melalui cross-validation tanpa overfitting, menunjukkan stabilitas performa pada data tes.

4. **Efisiensi**: Tanpa tuning, **XGBoost** mencapai performa yang sangat baik, memenuhi kebutuhan proyek tanpa meningkatkan beban komputasi.

**Kesimpulan Akhir**: **XGBoost** dipilih karena memberikan keseimbangan terbaik antara performa, efisiensi, dan kesesuaian dengan tujuan proyek, menjadikannya pilihan optimal untuk mendeteksi risiko default.

# Jadi berikut ini merupakan Model Terbaik: XGBoost Oversampled dengan Best Parameter (Setelah Hyperparameter)

### Table: XGBoost Metrics Before and After Hyperparameter Tuning (Oversampling)

| Metric                       | Before Tuning | After Tuning |
|------------------------------|---------------|--------------|
| Accuracy (Train Set)         | 0.9408        | 0.9460       |
| Accuracy (Test Set)          | 0.9390        | 0.9397       |
| Precision (Test Set)         | 0.9985        | 0.9966       |
| Recall (Test Set)            | 0.8183        | 0.8220       |
| F1-Score (Test Set)          | 0.8995        | 0.9009       |
| F2-Score (Test Set)          | 0.8489        | 0.8519       |
| ROC AUC (Train Set)          | 0.9647        | 0.9861       |
| ROC AUC (Test Set)           | 0.9471        | 0.9470       |
| Recall (Cross-Validation Train) | 0.8205     | 0.8239       |
| Recall (Cross-Validation Test)  | 0.8175     | 0.8208       |


### Analisa Model Terbaik: XGBoost Oversampled dengan Best Parameter

Berdasarkan tabel performa model sebelum dan setelah *hyperparameter tuning*, XGBoost dengan data oversampled menunjukkan peningkatan performa yang signifikan setelah tuning, menjadikannya model terbaik untuk prediksi risiko default:

1. **Peningkatan Akurasi**: Akurasi pada data *train* meningkat dari **94.08%** menjadi **94.60%**, menunjukkan model semakin mampu mempelajari pola data dengan baik tanpa *overfitting*. Akurasi pada data *test* tetap stabil di angka **93.97%**, menunjukkan kemampuan generalisasi yang baik.

2. **Precision yang Stabil**: Precision sedikit menurun dari **99.85%** menjadi **99.66%**, tetapi tetap sangat tinggi. Hal ini menunjukkan model tetap efisien dalam meminimalkan *false positives*.

3. **Peningkatan Recall**: Recall meningkat dari **81.83%** menjadi **82.20%**, yang sangat penting dalam konteks prediksi risiko default. Ini berarti model berhasil menangkap lebih banyak kasus default.

4. **F1-Score dan F2-Score yang Meningkat**: F1-Score naik dari **89.95%** menjadi **90.09%**, menunjukkan keseimbangan yang lebih baik antara precision dan recall. F2-Score, yang lebih menekankan recall, juga meningkat dari **84.89%** menjadi **85.19%**, mendukung tujuan utama meminimalkan risiko gagal bayar.

5. **ROC AUC yang Sangat Tinggi**: ROC AUC pada data *train* meningkat dari **96.47%** menjadi **98.61%**, mengindikasikan kemampuan model untuk membedakan antara nasabah default dan non-default semakin baik. ROC AUC pada data *test* tetap sangat tinggi di angka **94.70%**, menegaskan kualitas prediksi model.

6. **Stabilitas dalam Cross-Validation**: Recall pada *cross-validation* meningkat pada data *train* (dari **82.05%** ke **82.39%**) dan pada data *test* (dari **81.75%** ke **82.08%**), menunjukkan model yang lebih konsisten pada berbagai lipatan data.

### Kesimpulan:
XGBoost setelah *hyperparameter tuning* dengan data oversampled menjadi pilihan terbaik karena menunjukkan peningkatan yang signifikan dalam metrik utama, terutama recall, F1-score, dan ROC AUC. Model ini sangat cocok untuk memprediksi risiko default, memastikan jumlah kasus default yang lebih tinggi dapat diidentifikasi tanpa mengorbankan presisi.